src/ensembl/src/shared/components/upload/Upload.tsx

```diff
@@ -0,0 +1,193 @@
+import React, { useState } from 'react';
+import get from 'lodash/get';
+import classNames from 'classnames';
+
+import windowService from 'src/services/window-service';
+
+import styles from './Upload.scss';
+
+export enum FileReaderMethod {
+  ARRAY_BUFFER = 'readAsArrayBuffer',
+  BINARY_STRING = 'readAsBinaryString',
+  DATA_URL = 'readAsDataURL',
+  TEXT = 'readAsText'
+}
+type PropsForRespondingWithASingleFile = {
+  onChange: (file: File) => void;
+  callbackWithFiles: true;
+  allowMultiple: false;
+};
+
+type PropsForRespondingWithMultipleFiles = {
+  onChange: (files: FileList) => void;
+  callbackWithFiles: true;
+  allowMultiple: true;
+};
+
+type PropsForRespondingWithContent = {
+  onChange: (contents: string[]) => void;
+  callbackWithFiles: false;
+  fileReaderMethod: FileReaderMethod;
+};
+
+type OnChangeProps =
+  | PropsForRespondingWithASingleFile
+  | PropsForRespondingWithMultipleFiles
+  | PropsForRespondingWithContent;
+
+export type UploadProps = {
+  id: string;
+  name?: string;
+  label: string;
+  classNames?: {
+    default?: string;
+    active?: string;
+  };
+  allowMultiple?: boolean;
+} & OnChangeProps;
+
+const Upload = (props: UploadProps) => {
+  const [drag, setDrag] = useState(false);
+
+  const fileReaders: FileReader[] = [];
+  let totalPendingFilesToRead = 0;
+
+  const handleDrag = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+  };
+  const handleDragIn = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    if (get(e, 'dataTransfer.items.length', 0) > 0) {
+      setDrag(true);
+    }
+  };
+
+  const handleDragOut = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    setDrag(false);
+  };
+
+  const handleFileRead = () => {
+    totalPendingFilesToRead--;
+    // Do not return the content until all files are read
+    if (totalPendingFilesToRead > 0 || props.callbackWithFiles) {
+      return;
+    }
+
+    const contents: string[] = fileReaders.map(
+      (fileReader) => fileReader.result as string
+    );
+
+    props.onChange(contents);
+  };
+
+  const handleDrop = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    setDrag(false);
+
+    if (
+      e.dataTransfer &&
+      e.dataTransfer.files &&
+      e.dataTransfer.files.length > 0
+    ) {
+      if (props.callbackWithFiles) {
+        if (!props.allowMultiple) {
+          props.onChange(e.dataTransfer.files[0]);
+          return;
+        }
+        props.onChange(e.dataTransfer.files);
+        return;
+      }
+
+      const { files } = e.dataTransfer;
+
+      [...files].forEach((file) => {
+        const fileReader = windowService.getFileReader();
+        fileReaders.push(fileReader);
+        totalPendingFilesToRead++;
+        fileReader.onloadend = handleFileRead;
+        fileReader[props.fileReaderMethod](file);
+      });
```

Comment:
Two questions regarding this:

1) Technical — wouldn't it be nicer, instead of having the totalPendingFilesToRead and fileReaders variables floating around and being mutated by various functions, to re-write this bit using promises:

```
const readPromises = [...files].map((file) => {
  return new Promise((resolve, reject) => {
    const fileReader = windowService.getFileReader();
    fileReaders.push(fileReader);
    totalPendingFilesToRead++;
    fileReader.onloadend = resolve(fileReader);
    fileReader[props.fileReaderMethod](file);

    // do we handle error?
  });
});

Promise.all(readPromises).then(handleFiles); // <- this will pass all filereaders to the handleFiles function
```

2) How should the code behave if not all files have been successfully read? Your implementation will probably just hang there (since `totalPendingFilesToRead` will be greater than 0; and the above implementation with Promise.all will abort all promises (meaning that for the promise implementation each individual promise going into Promise all should have an individual catch).

Reply:
> re-write this bit using promises

I will look into implementing it using promises.

> How should the code behave if not all files have been successfully read? 
 I don't think it will hang because `onloadend` will be called even if there is a failure. I am still not sure what should happen when there is a failure as there are many scenarios we need to consider. 

I am also not sure how likely this is to happen but some the possible solutions would be:
1. Display an error below the drop-zone mentioning the file name and cancel the process.
2. Proceed with the remaining files and rely on the users to validate if all the file contents are read.
3.  Add an `error` parameter to the onChange callback and get the errors handled by the parent. (I would prefer this)

Reply:
> do we handle error?

I am starting to think if we really need error handling as the chances are very less for an error to occur while selecting a file.

Do you know any way this error can be triggered?